### PR TITLE
Calculate proper code line number from added lines

### PR DESCRIPTION
**Pull request by Genki-S**
Problem: added lines cannot be used to map changed code to test files.

Since added lines did not exist in previous file, the line number of those cannot be used to determine which tests to run (since test-to-code mapping is a mapping from line number in **previous file** to a test file).

For example, if there is a mapping like "only line 10 of foo.rb should run test file foo_test.rb", if a change introduced new 10 blank lines at the top of foo.rb, current implementation selects foo_test.rb even if line 20 of new foo.rb (i.e. line 10 of old foo.rb) is not changed. I changed my test case to reflect similar case: https://github.com/Genki-S/ttnt/commit/45332c6952d81482114829ddc2604daca935c633 Explanation follows:

This test case produces git diff like this (in side-by-side diff):
```
diff --git c/lib/fizzbuzz.rb w/lib/fizzbuzz.rb
index fda1356..87f7ddf 100644
--- c/lib/fizzbuzz.rb
+++ w/lib/fizzbuzz.rb
@@ -1,3 +1,13 @@
                                                                                     1
                                                                                     2
                                                                                     3
                                                                                     4
                                                                                     5
                                                                                     6
                                                                                     7
                                                                                     8
                                                                                     9
                                                                                    10
 1 # This is meaningless comment to produce complex git diff:                       11 # This is meaningless comment to produce complex git diff:
 2 # Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod       12 # Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
 3 # tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,   13 # tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,
@@ -10,7 +20,7 @@ def fizzbuzz_convert(n)
10   if n % 15 == 0                                                                 20   if n % 15 == 0
11     "fizzbuzz"                                                                   21     "fizzbuzz"
12   elsif n % 3 == 0                                                               22   elsif n % 3 == 0
13     "fizz"                                                                       23     "foo"
14   elsif n % 5 == 0                                                               24   elsif n % 5 == 0
15     "buzz"                                                                       25     "buzz"
16   else                                                                           26   else
```

Line 13 of old fizzbuzz.rb is changed from "fizz" to "foo", as line 23 of new fizzbuzz.rb. However, since 10 blank lines are added to the top of new fizzbuzz.rb, `line.new_lineno` of the changed line (line of "foo") becomes 23. The line number 23 is meaningless since the old fizzbuzz.rb does not even contain the line 23. So, current implementation fails on some tests on the commit: https://github.com/Genki-S/ttnt/commit/45332c6952d81482114829ddc2604daca935c633

So, this PR makes it possible to deal with such cases by not using added lines to select test cases. Instead, I changed to select tests using a line **before** consecutive added lines (which, I think, is always the type of "context" or "deletion"). Since the "context" or "deletion" lines have `old_lineno` attribute, it can be used to select tests. (As for an edge case when lines are added in the top of a hunk, like when lines are added on top of a file, I used `hunk.old_start` instead).

In a nutshell, we should always use `line.old_lineno` and should not use `line.new_lineno`, since test-to-code mapping is based on line numbers of old files.

@robin850 could you review the code please? If my explanation is not clear, please let me know :pray:

### CODE REVIEW

File: test/test_selector_test.rb [hunk 1]
```diff
@@ -17,9 +17,9 @@ def test_base_obj_selection
     end
 
     def test_selects_tests
-      assert_equal @selector.tests, nil
-      assert_equal @selector.select_tests!.to_a, ['test/fizz_test.rb']
-      assert_equal @selector.tests.to_a, ['test/fizz_test.rb']
+      assert_equal nil, @selector.tests
+      assert_equal ['test/fizz_test.rb'], @selector.select_tests!.to_a
+      assert_equal ['test/fizz_test.rb'], @selector.tests.to_a
```
Comment (Genki-S) [inline test/test_selector_test.rb]: I'm sorry I accidentally included changes not related to this PR :sweat:

**Comment by robin850** [comment]
Hi Genki!

Good job! Just a tiny comment though: do you need to both add dummy comments and do the `"\n" * 10` thingy ? Either one should be fine I guess.

Sorry it took me a while to answer ; actually the explanation is clear but I didn't understand the purpose of the patch first. :smile:

:+1: for the other changes though, even if they're not directly related.

**Comment by robin850** [comment]
Hum, actually I'm wrong, the comments are pretty useless ; the `"\n" * 10` thingy is what will expose the problem you're trying to solve, no ?

**Comment by Genki-S** [comment]
Yes, you are right! :bulb: Thank you for mentioning it, I didn't notice...
My intention was to split hunks to 2, but it was meaningless anyway.

I noticed one bug which adds `hunk.old_start` in unintended cases, so I fixed it: https://github.com/Genki-S/ttnt/commit/710ab69f537bbadf7d28f21f8c48fff34aba90ac

I dealt with removing comments in https://github.com/Genki-S/ttnt/commit/4d9fe180a45aa9539052fadecf069f2a95675aea and https://github.com/Genki-S/ttnt/commit/9d8c36286c6c3d3f9e9bc3f18ee56049a9427e89, the second commit is needed because the first line of old `fizzbuzz.rb` (which is the line of `def fizzbuzz(n)`) is selected as a target line in `TestSelector#select_tests_from_patch`, and since `test_fizz.rb` and `test_buzz.rb` both executes the line, some tests (like `TestSelectorTest#test_selects_tests`) did not work well. (This requires too much knowledge for fixture, so I want to deal with it somehow... but I have no good idea for now :cry: )

**Comment by robin850** [comment]
Could you add a regression test for the fix you addressed in https://github.com/Genki-S/ttnt/commit/710ab69f537bbadf7d28f21f8c48fff34aba90ac please ? :-)

> This requires too much knowledge for fixture, so I want to deal with it somehow... but I have no good idea for now

What is the problem with you current approach of ugglyfying elsewhere ?

**Comment by Genki-S** [comment]
Actually, 2 existing tests fails like this without the change in https://github.com/Genki-S/ttnt/commit/710ab69f537bbadf7d28f21f8c48fff34aba90ac :

```
  1) Failure:
TTNT::TestSelectorTest#test_selects_tests [/Users/genki/dev/ttnt/test/test_selector_test.rb:20]:
--- expected
+++ actual
@@ -1 +1 @@
-["test/fizz_test.rb"]
+["test/buzz_test.rb", "test/fizz_test.rb"]



  2) Failure:
TTNT::IntegrationTest#test_fizz_test_is_selected [/Users/genki/dev/ttnt/test/integration_test.rb:28]:
Expected /1\ runs,\ 1\ assertions,\ 1\ failures/ to match "Run options: --seed 10729\n\n# Running:\n\n.F\n\nFinished in 0.001193s, 1676.3476 runs/s, 3352.6951 assertions/s.\n\n  1) Failure:\nTestFizz#test_fizz [/private/var/folders/qw/2lhv57dj6yn835dvyk5x2qn40000gn/T/ttnt_repository20150719-20316-40uhuw/test/fizz_test.rb:6]:\nExpected: \"fizz\"\n  Actual: \"foo\"\n\n2 runs, 4 assertions, 1 failures, 0 errors, 0 skips\n".
```

This is because of the change in fixture file `fizzbuzz.rb` (https://github.com/Genki-S/ttnt/commit/4d9fe180a45aa9539052fadecf069f2a95675aea), which made the first line `def fizzbuzz_convert(n)`, and the bug fixed in https://github.com/Genki-S/ttnt/commit/710ab69f537bbadf7d28f21f8c48fff34aba90ac was selecting the 1st line as target line. Thus those test fails.

In conclusion, the bug was found after the change of https://github.com/Genki-S/ttnt/commit/4d9fe180a45aa9539052fadecf069f2a95675aea which made above 2 tests fail, and I fixed it in https://github.com/Genki-S/ttnt/commit/710ab69f537bbadf7d28f21f8c48fff34aba90ac, I'm so sorry the order of commits is wrong and it was confusing... :cry:

Are those tests enough for regression? (I tried to write another test, but the contents was no different from `TTNT::TestSelectorTest#test_selects_tests`)

> What is the problem with you current approach of ugglyfying elsewhere ?

I thought this was a little confusing since the place you add lines influences the test results. But thinking again, it's natural how you change your fixture influence the test results. So now I'm fine with it, thank you for the question :smile:

**Comment by robin850** [comment]
> I'm so sorry the order of commits is wrong and it was confusing

Ah ok, I see, no problem!

If you want to reorganize commits order, you can use the `git rebase` command. Just use `git rebase HEAD~n` if you want to change the order of the n last commits ; you just have to change the order of the lines to change the order of the commits. You can `cp` you current directory if you want to experiment first! :-)

Sometimes it's not that easy because there may be kind of "anachronisms" like if you have commit 1 and 2, with commit 2's diff integrating changes of commit 1 and you try to put 2 before 1 in history. You can e-mail me questions about Git if you have problems and I will happily answer!

Then I guess it's fine, you can merge this!

**Comment by Genki-S** [comment]
> If you want to reorganize commits order, you can use the git rebase command

Oh yes! I couldn't come up with it :sweat: Thank you!

> You can e-mail me questions about Git if you have problems and I will happily answer!

Thank you so much! :smile: This time, luckily I could reorder the commits without any problem :dancer: 

> Then I guess it's fine, you can merge this!

Thank you! :tada:
